### ISSUE TITLE
Revoked sessions are not beeing revoked.

### ISSUE THREAD

**Issue by dtorrero** [mastodon/mastodon#31631]
### Steps to reproduce the problem

1. Log in with android client Tusky.

2. Log in with the same account on a desktop browser and go to Settings -> Account -> Revoke session, and revoke the recently created session.

3. Go back to the Tusky session and  interact with the app, watch content, post content.

### Expected behaviour

The session should log off when it is revoked.

### Actual behaviour

Session stays logged in.

### Detailed description

When revoking a session the revoked session has still full access to the session, can still see, and toot content. 

This can be reproduced using the android Tusky client for the revoked session. 

How to reproduce: 

From desktop browser mastodon client in Settings -> Account -> Revoke session  

If the client is using a browser version of the client, the client will be signed off and not allowed to do any actions. If the client is Tusky, there will be no effect on the session, the user can still interact normally with that session. 

Also changing the password or enabling the Two-factor Auth has no effect on this session either. 

### Mastodon instance

metalhead.club

### Mastodon version

 v4.2.12

**Comment by daprice** [comment]
Are you talking about revoking it from Settings > Account, or Settings > Account > Authorized apps?

The list of sessions in Settings > Account says "These are the web browsers currently logged in to your Mastodon account” so I don’t think apps are supposed to show up there.

**Comment by dtorrero** [comment]
You are right, removing them from Authorized apps does revoke the sessions. The confusing part, as you said, is that you can see the app session in the list with the browser sessions, and revoking it there has no effect.  Thanks for your answer.

**Comment by ThisIsMissEm** [comment]
So tusky likely uses a web view for having you log in to your account, that's why you see the web session, and why revoking it appears to not do anything — after that initial authentication in the web view that session is discarded and access tokens & authorized apps take over.

This leads to two questions:
1. Should revoking the web session which authorized an OAuth application also revoke that application's access tokens?
2. Should we link to authorized applications on the page where we show sessions?

**Comment by renchap** [comment]
Another possibility: close the web session once the client got the access tokens, as it is no longer needed

**Comment by ThisIsMissEm** [comment]
> Another possibility: close the web session once the client got the access tokens, as it is no longer needed

Apps would need to handle the logic for making the logout request within the web view.

Having OAuth do that wouldn't be a solution, because then you'd be logged out of Web UI every time you use an OAuth App

**Comment by daprice** [comment]
On iOS it’s common to use the regular browsing session for in-app authorization, so you wouldn’t want the session to be closed afterward. Some apps/situations do use an ephemeral browsing session (i.e. the session would only be used for that one authorization attempt), but the app has no way of making the logout request (that I know of) because it’s using the system web browser and the app has no control of it. For good reason, because doing auth in an app-controlled web view gives the app the ability to read login credentials (if that’s what Tusky is doing… yikes. But maybe there isn’t a better solution on android 🤷🏻‍♂️).

The only way I can think of to distinguish between these on the server side would be to add a parameter to `authorize` (call it `isEphemeralSession` or something) to specify that the session should be closed when the authorization attempt finishes. But that still relies on each app setting it properly.

**Comment by ThisIsMissEm** [comment]
> The only way I can think of to distinguish between these on the server side would be to add a parameter to authorize

we shouldn't be extending OAuth in non-standard ways, or at least not without an I-D at minimum (like #31151)

i think just associating the OAuth Access Tokens with the session that created them would be the way to go, and then give an option of revoking authorized applications when revoking sessions

**Comment by daprice** [comment]
That and put the authorized apps and browser sessions on the same page. From a user perspective I think it makes sense to have a page of “here’s all the things with access to your account” rather than having them separate. At the very least, they should link to each other like you suggested earlier.

**Comment by fowl2** [comment]
So the problem here is a client app is redirecting though the web browser when it receives a token error, and since the web browser is logged in it's silently issuing a new one?

I'm pretty sure I've seen other platforms use an interstitial/confirmation page when logging into an app via the web. 

Couldn't we do that for *every* token acquisition? 

----
You're logging in to *account* on with *app name*. {View / revoke previously granted permissions link}

This will replace {session details}. 

[Continue] [Switch Accounts]

**Comment by ThisIsMissEm** [comment]
@fowl2 we already have something like that but that is not the issue here: the app has it's own set of cookies separate from the native browser on the platform, it's using a webview or similar to perform the login with the user-agent of "Tusky". The act of logging in creates a SessionActivation, this happens for **all** logins regardless of where they come from.

For many apps, they delegate to either the system browser or the system framework for doing OAuth, so those create a "iOS" or "Android" session.

So when you look at the list of Sessions you're seeing that SessionActivation for Tusky, and there's a revoke button, so people assume revoking it would revoke the OAuth Access Token that Tusky requests.

OAuth Access Tokens (authorized applications) are separate from browser sessions, there's no link between them presently.

**Comment by mjankowski** [comment]
Putting aside the more complex scenario of apps closing out their own temporary browser sessions or something similar - is there an app copy/hints/instructions change we could make to better clarify the difference between the authorized web sessions vs authorized apps?

Could be as simple as adding a link/reference to the other area from each area (ie, from web sessions - "There are also [authorized applications](#apps) to review" or whatever, and vice versa). Does not solve the more nuanced issues, but might add clarity for basic confusion scenario.
